libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,575 @@
+       return c.count(*f) != 0;
+   }
+};
+// Field-not-in-list is just field-in-list wrapped in a negator
+template<typename Field, typename Container> struct predicate_not_in : predicate_in<Field, Container> {
+   using base = predicate_in<Field, Container>;
+   bool operator()(const Field& f, const Container& c) const { return !base::operator()(f, c); }
+};
+
+// List-contains-list predicate
+template<typename C1, typename C2, typename = void> struct predicate_has_all : predicate_invalid<C1, C2> {};
+template<typename FieldElement, typename ArgumentElement>
+struct predicate_has_all<flat_set<FieldElement>, flat_set<ArgumentElement>,
+                         std::enable_if_t<comparable_types<FieldElement, ArgumentElement>>> {
+   // Field is already flat_set
+   constexpr static bool valid = true;
+   bool operator()(const flat_set<FieldElement>& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      return std::includes(f.begin(), f.end(), a.begin(), a.end());
+   }
+};
+template<typename FieldContainer, typename ArgumentElement>
+struct predicate_has_all<FieldContainer, flat_set<ArgumentElement>,
+                         std::enable_if_t<is_container<FieldContainer> && !is_flat_set<FieldContainer> &&
+                                          comparable_types<typename FieldContainer::value_type, ArgumentElement>>> {
+   predicate_has_all<flat_set<typename FieldContainer::value_type>, flat_set<ArgumentElement>> inner;
+   // Field is other container; convert to flat_set
+   constexpr static bool valid = true;
+   bool operator()(const FieldContainer& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      flat_set<typename FieldContainer::value_type> fs(f.begin(), f.end());
+      return inner(fs, a);
+   }
+};
+template<typename OptionalType, typename Argument>
+struct predicate_has_all<fc::optional<OptionalType>, Argument, void> : predicate_has_all<OptionalType, Argument> {
+   // Field is optional container
+   bool operator()(const fc::optional<OptionalType>& f, const Argument& a) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+      return (*this)(*f, a);
+   }
+};
+
+// List contains none of list predicate
+template<typename C1, typename C2, typename = void> struct predicate_has_none : predicate_invalid<C1, C2> {};
+template<typename FieldElement, typename ArgumentElement>
+struct predicate_has_none<flat_set<FieldElement>, flat_set<ArgumentElement>,
+                          std::enable_if_t<comparable_types<FieldElement, ArgumentElement>>> {
+   // Field is already flat_set
+   constexpr static bool valid = true;
+   bool operator()(const flat_set<FieldElement>& f, const flat_set<ArgumentElement>& a) const {
+      flat_set<FieldElement> intersection;
+      std::set_intersection(f.begin(), f.end(), a.begin(), a.end(),
```

Comment:
Computing the intersection only to see if it is empty may be more expensive than required.

Reply:
True... OK, how's this instead?
```cpp
return !std::any_of(f.begin(), f.end(), [&a](const FieldElement& fe) { return a.count(fe) > 0; });
```

Reply:
Actually that's worse, because it does not make use of the fact that both `f` and `a` are sorted.
`std::set_intersection` already has linear complexity, only it doesn't stop on the first hit. The new version has `O(n*log(m))` in the worst case.

Reply:
Very true... Hmm, so maybe I should just keep the `set_intersection` version, but remove the `intersection` set and make a fake insertion iterator that throws when you attempt to use it.

Reply:
Throwing is nasty because I'd be catching in the same function and returning false... but it does work. :P

Reply:
Perhaps it's best to keep `std::set_intersection` for now. https://news.ycombinator.com/item?id=9196301

Reply:
True words. :P The fake iterator version is simple enough too, but I'll leave it out for now.